### PR TITLE
Use dispatch_to_series where possible

### PR DESCRIPTION
**Pull request by jbrockmendel**
Broken off from #22534, since that ended up surfacing a bug elsewhere.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -4806,13 +4806,14 @@ def _arith_op(left, right):
...
-        return self._constructor(result, index=new_index, columns=new_columns,
-                                 copy=False)
+            return self._constructor(result,
+                                     index=new_index, columns=new_columns,
+                                     copy=False)
 
     def _combine_match_index(self, other, func, level=None):
         left, right = self.align(other, join='outer', axis=0, level=level,
                                  copy=False)
+        assert left.index.equals(right.index)
```
Comment (gfyoung) [inline pandas/core/frame.py]: Why do we need these assert statements?  Are they tested anywhere?

Comment (jbrockmendel) [reply pandas/core/frame.py]: These assertions are always true.  I added these so I don't have to keep looking back at core.ops (the only place from which these are called) to double-check what's going on.

Comment (gfyoung) [reply pandas/core/frame.py]: So this is just a sanity check really?  Perhaps we should comment that then?

File: pandas/core/ops.py [hunk 2]
```diff
@@ -1638,6 +1638,7 @@ def dispatch_to_series(left, right, func):
     """
     # Note: we use iloc to access columns for compat with cases
     #       with non-unique columns.
+    right = lib.item_from_zerodim(right)
```
Comment (jreback) [inline pandas/core/ops.py]: is there a reason you needed to add this? (e.g. do we actually hit this in the new code path and not in the original)?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/ops.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/ops.py]: Yes.  This PR has `DataFrame._combine_const` send zero-dim arrays down this path, which without this line don't get handled correctly.

Comment (jreback) [reply pandas/core/ops.py]: ok great

**Review by jreback — APPROVED** [review]

**Pull request merged**